File: src/main/java/com/antwerkz/ophelia/utils/MongoUtil.java
package com.antwerkz.ophelia.utils;

import com.antwerkz.ophelia.models.MongoCommand;
import com.antwerkz.sofia.Ophelia;
import com.mongodb.DBCollection;
import com.mongodb.DBCursor;
import com.mongodb.DBObject;
import com.mongodb.MongoClient;
import com.mongodb.WriteResult;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class MongoUtil {
    private MongoClient client;

    public MongoUtil(final MongoClient client) {

        this.client = client;
    }

    public List<Map> explain(MongoCommand command) {
        return Arrays.asList(getCursor(command).explain().toMap());
    }

    public List<Map> query(MongoCommand command) {
        DBCursor query = getCursor(command);
        query.sort(command.getSortDocument());
        query.limit(command.getLimit());
        return extract(query);
    }

    public List<Map> insert(final MongoCommand command) {
        DBCollection collection = client.getDB(command.getDatabase()).getCollection(command.getCollection());
        return wrap(collection.insert(command.getInsertDocument()).getN());
    }

    public List<Map> update(final MongoCommand command) {
        DBCollection collection = client.getDB(command.getDatabase()).getCollection(command.getCollection());
        WriteResult update = collection.update(command.getQueryDocument(),
                                               command.getUpdateDocument(),
                                               command.getUpsert(),
                                               command.getMultiple());
        return wrap(update.getN());
    }
    public List<Map> remove(final MongoCommand command) {
        DBCollection collection = client.getDB(command.getDatabase()).getCollection(command.getCollection());
        return wrap(collection.remove(command.getQueryDocument()).getN());
    }

    private DBCursor getCursor(final MongoCommand command) {
        DBCollection collection = client.getDB(command.getDatabase()).getCollection(command.getCollection());
        return collection.find(command.getQueryDocument(), command.getProjectionsDocument());
    }

    @SuppressWarnings("unchecked")
    private List<Map> count(final DBCollection collection) {
        Map map = new HashMap();
        map.put(Ophelia.count(), collection.count(null));
        return Arrays.asList(map);
    }


    private Object extract(final DBObject eval, final String first, final String second) {
        return ((DBObject) eval.get(first)).get(second);
    }

    private List<Map> extract(DBCursor execute) {
        List<Map> list = new ArrayList<>();
        try (DBCursor cursor = execute) {
            for (final DBObject dbObject : cursor) {
                list.add(dbObject.toMap());
            }
            if (list.isEmpty()) {
                Map<String, String> map = new TreeMap<>();
                map.put("message", Ophelia.noResults());
                list.add(map);
            }
        }
        return list;
    }

    private List<Map> wrap(int number) {
        Map<String, Number> count = new TreeMap<>();
        count.put("count", number);
        return Arrays.<Map>asList(count);
    }
}

File: src/main/kotlin/com/antwerkz/ophelia/utils/MongoUtil.kt
package com.antwerkz.ophelia.utils

import com.antwerkz.ophelia.models.MongoCommand
import com.antwerkz.sofia.Ophelia
import com.fasterxml.jackson.databind.ObjectMapper
import com.mongodb.BasicDBObject
import com.mongodb.MongoClient
import com.mongodb.client.FindIterable
import com.mongodb.client.model.UpdateOptions
import org.bson.Document
import java.util.ArrayList
import java.util.Arrays

public class MongoUtil(private val client: MongoClient) {

    public fun explain(command: MongoCommand): List<Document> {
        val explain = client.getDB(command.database)
                .getCollection(command.collection)
                .find(BasicDBObject(command.getQueryDocument()), BasicDBObject(command.getProjectionsDocument().toLinkedMap()))
                .explain()
                as BasicDBObject
        val document = Document()
        explain.entrySet()
                .forEach { document.put(it.key : String, it.value) }
        return Arrays.asList<Document>(document)
    }

    public fun query(command: MongoCommand): List<Document> {
        val query = getCursor(command)
        query.sort(command.getSortDocument())
        query.limit(command.limit)
        return extract(query)
    }

    public fun insert(command: MongoCommand) {
        val collection = client.getDatabase(command.database).getCollection(command.collection)
        return collection.insertOne(command.getInsertDocument())
    }

    public fun update(command: MongoCommand): Long {
        val collection = client.getDatabase(command.database).getCollection(command.collection)
        val upsert = command.upsert
        return if (command.multiple) {
            collection.updateOne(command.getQueryDocument(), command.getUpdateDocument(), UpdateOptions().upsert(upsert))
                    .getModifiedCount()
        } else {
            collection.updateMany(command.getQueryDocument(), command.getUpdateDocument(), UpdateOptions().upsert(upsert))
                    .getModifiedCount()
        }
    }

    public fun remove(command: MongoCommand): Long {
        val collection = client.getDatabase(command.database).getCollection(command.collection)
        val remove = collection.deleteMany(command.getQueryDocument())
        return remove.getDeletedCount()
    }

    public fun count(command: MongoCommand): Int {
        val collection = client.getDatabase(command.database).getCollection(command.collection)
        return collection.find(command.getQueryDocument()).count()
    }

    private fun getCursor(command: MongoCommand): FindIterable<Document> {
        return client.getDatabase(command.database)
                .getCollection(command.collection)
                .find(command.getQueryDocument())
                .projection(command.getProjectionsDocument())
    }


    private fun extract(execute: FindIterable<Document>): List<Document> {
        val list = ArrayList<Document>()
        execute.forEach { document ->
            list.add(document)
        }
        if (list.isEmpty()) {
            val map = Document()
            map.put("message", Ophelia().noResults())
            list.add(map)
        }
        return list
    }
}

fun Document.pretty(): String {
    return ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(this);
}